examples/multi-threaded/main.c

```diff
@@ -24,79 +31,73 @@ static void start_thread(void (*f)(void *), void *p) {
 #endif
 }
 
-struct userdata {
-  int sock;  // Paired socket, worker thread owns it
-  struct mg_str
-      body;  // data to be processed
-};
+enum { TO_WORKER = '>', FROM_WORKER = '<' };  // Marks message direction
 
-static void thread_function(void *param) {
-  struct userdata *p = (struct userdata *) param;
-  sleep(2);  // Simulate long execution
-  // Respond, wakeup event manager
-  if (p->body.len)
-    send(p->sock, p->body.ptr, p->body.len, 0);  // if there is a body, echo it
-  else
-    send(p->sock, "hi", 2, 0);  // otherwise just respond "hi"
-  close(p->sock);               // Close the connection
-  free((void *) p->body.ptr);   // free body
-  free(p);                      // free userdata
-}
+static void worker_thread(void *param) {
+  struct mg_queue *q = (struct mg_queue *) param;  // A queue - we own it
+  struct mg_str name;
+  unsigned long len, ofs;
 
-static void link_conns(struct mg_connection *c1, struct mg_connection *c2) {
-  c1->fn_data = c2;
-  c2->fn_data = c1;
-}
+  usleep(200 * 1000);  // Simulate some delay, 200ms
 
-static void unlink_conns(struct mg_connection *c1, struct mg_connection *c2) {
-  c1->fn_data = c2->fn_data = NULL;
-}
+  // Wait for a message, read it and make a copy of it
+  while ((len = mg_qnext(q, &ofs)) == MG_QUEUE_EMPTY) usleep(1000);  // Wait
+  name = mg_strdup(mg_str_n((char *) q->buf + ofs + 1, len - 1));    // Copy
+  mg_qconsume(q);                            // Delete received message
+  MG_INFO(("Got message: [%s]", name.ptr));  // Show what we've got
 
-// Pipe event handler
-static void pcb(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
-  struct mg_connection *parent = (struct mg_connection *) fn_data;
-  MG_INFO(("%lu %p %d %p", c->id, c->fd, ev, parent));
-  if (parent == NULL) {  // If parent connection closed, close too
-    c->is_closing = 1;
-  } else if (ev == MG_EV_READ) {  // Got data from the worker thread
-    mg_http_reply(parent, 200, "Host: foo.com\r\n", "%.*s\n", c->recv.len,
-                  c->recv.buf);  // Respond!
-    c->recv.len = 0;             // Tell Mongoose we've consumed data
-  } else if (ev == MG_EV_OPEN) {
-    link_conns(c, parent);
-  } else if (ev == MG_EV_CLOSE) {
-    unlink_conns(c, parent);
-  }
+  // Write a response back
+  usleep(1 * 1000 * 1000);  // Simulate long execution
+  while ((len = mg_qspace(q, &ofs)) < name.len + 10) usleep(1000);  // Wait
+  MG_INFO(("Got space %lu in the queue, offset %lu", len, ofs));
+  len = mg_snprintf((char *) q->buf + ofs, len, "%cHello %s", FROM_WORKER,
+                    name.ptr);  // Copy message, prepend direction marker
+  mg_qproduce(q, ofs, len);     // Signal that we've written a message
+  free((char *) name.ptr);      // Free stored name
+
+  // Wait until connection reads the message, then deallocate the queue
+  while ((len = mg_qnext(q, &ofs)) != MG_QUEUE_EMPTY) usleep(1000);
+  MG_INFO(("done, deallocating queue"));
+  free(q);
```

Comment:
I'd rather free() where I calloc(), what would a real life application for this ? Does it make sense that the worker waits for the master to have consumed the response to release the queue ? I would kill the worker task asap and let the master decide whether to reuse the queue or dealloc.

Reply:
Got it